File: js/src/sorting/Quick3Way.js
import Sorter from './Sorter';

class Quick3Way extends Sorter {
    constructor(arr = []) {
        super(arr);
    }

    sort() {
        let N = this.arr.length;
        this.sortRec(this.arr, 0, N - 1);
    }

    sortRec(arr, lo, hi) {
        if (hi <= lo) return;
        let lt = lo;
        let i = lo + 1;
        let gt = hi;
        let v = arr[lo];
        while (i <= gt) {
            if (arr[i] < v) this.exch(i++, lt++);
            else if (arr[i] > v) this.exch(i, gt--);
            else i++;

        }

        this.sortRec(arr, lo, lt - 1);
        this.sortRec(arr, gt + 1, hi);
    }
}

module.exports = Quick3Way;

File: java/src/com/test/sorting/Quick3Way.java
package com.test.sorting;

public class Quick3Way extends Sorter
{
    public Quick3Way(int[] arr)
    {
        super(arr);
    }

    @Override
    public void sort()
    {
        int N = arr.length;
        sort(arr, 0, N - 1);
    }

    private void sort(int[] arr, int lo, int hi)
    {
        if (hi <= lo) return;
        int lt = lo, i = lo + 1, gt = hi;
        int v = arr[lo];
        while (i <= gt) {
            if (arr[i] < v) exch(lt++, i++);
            else if (arr[i] > v) exch(i, gt--);
            else i++;
        }

        sort(arr, lo, lt -1);
        sort(arr, gt + 1, hi);
    }
}
